### PR TITLE
Optimize trigger

### PR DESCRIPTION
**Pull request by nearnshaw**

### CODE REVIEW

File: src/trigger.ts [hunk 1]
```diff
@@ -15,422 +15,454 @@ export const ALL_LAYERS = 255
...
+					}
+				}
+			}
+		}
+
+		return false
+	}
+
+	function computeCollisions(entity: Entity, shapeWorldPos: Map<Entity, Array<Vector3>>) {
+		let collisions: Set<Entity> = new Set()
```
Comment (wacaine) [inline src/trigger.ts]: assign to EMPTY_IMMUTABLE_SET

File: src/trigger.ts [hunk 2]
```diff
@@ -15,422 +15,454 @@ export const ALL_LAYERS = 255
...
+
+				if (!otherTrigger.active)
+					continue
+
+				if (!(trigger.triggeredByMask & otherTrigger.layerMask))
+					continue
+
+				const intersecting = areTriggersIntersecting(shapeWorldPos.get(entity)!, trigger, shapeWorldPos.get(otherEntity)!, otherTrigger)
+				if (intersecting)
+					collisions.add(otherEntity)
```
Comment (wacaine) [inline src/trigger.ts:220]: needs check if (collisions === EMPTY_IMMUTABLE_SET) collisions = new Set()

File: src/trigger.ts [hunk 3]
```diff
@@ -15,422 +15,454 @@ export const ALL_LAYERS = 255
...
+
+	function computeCollisions(entity: Entity, shapeWorldPos: Map<Entity, Array<Vector3>>) {
+		let collisions: Set<Entity> = new Set()
+		const trigger = Trigger.get(entity)
+
+		if (!trigger.active)
+			return collisions
+
+
+		if (trigger.triggeredByMask == 1) {
```
Comment (wacaine) [inline src/trigger.ts]: this is backwards breaking if someone registered other triggers as layer 1 to interact with layer 1.  "player multi ball pi machine"   make it 2.0 when release explaining change?

File: src/trigger.ts [hunk 4]
```diff
@@ -213,13 +215,15 @@ function createTriggers(targetEngine: IEngine) {
 					continue
 
 				const intersecting = areTriggersIntersecting(shapeWorldPos.get(entity)!, trigger, shapeWorldPos.get(otherEntity)!, otherTrigger)
-				if (intersecting)
-					collisions.add(otherEntity)
+				if (intersecting) { }
```
Comment (wacaine) [inline src/trigger.ts]: @nearnshaw  it should read.  you open then closed the if immediately :)


```
if (intersecting) {
    if (collisions === EMPTY_IMMUTABLE_SET) collisions = new Set()
    collisions.add(otherEntity)
}
```

Comment (nearnshaw) [reply src/trigger.ts]: haha yikes, I must have messed it up while refactoring
good catch!

**Pull request merged**